flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -0,0 +1,325 @@
+				restPort = getServiceNodePort(service, RestOptions.PORT);
+			}
+		} else if (service.getSpec().getExternalIPs() != null && service.getSpec().getExternalIPs().size() > 0) {
+			address = service.getSpec().getExternalIPs().get(0);
+		}
+		return new Endpoint(address, restPort);
+	}
+
+	@Override
+	public void stopAndCleanupCluster(String clusterId) {
+		this.internalClient.services().inNamespace(this.nameSpace).withName(clusterId).cascading(true).delete();
+	}
+
+	@Override
+	public void logException(Exception e) {
+		LOG.error("Kubernetes Exception: {}", e);
+	}
+
+	@Override
+	@Nullable
+	public FlinkService getInternalService(String clusterId) {
+		return getService(clusterId);
+	}
+
+	@Override
+	@Nullable
+	public FlinkService getRestService(String clusterId) {
+		return getService(clusterId + Constants.FLINK_REST_SERVICE_SUFFIX);
+	}
+
+	@Override
+	public void watchPodsAndDoCallback(Map<String, String> labels, PodCallbackHandler callbackHandler) {
+		Watcher<Pod> watcher = new Watcher<Pod>() {
+			@Override
+			public void eventReceived(Action action, Pod pod) {
+				switch (action) {
+					case ADDED:
+						callbackHandler.onAdded(Collections.singletonList(new FlinkPod(flinkConfig, pod)));
+						break;
+					case MODIFIED:
+						callbackHandler.onModified(Collections.singletonList(new FlinkPod(flinkConfig, pod)));
+						break;
+					case ERROR:
+						callbackHandler.onError(Collections.singletonList(new FlinkPod(flinkConfig, pod)));
+						break;
+					case DELETED:
+						callbackHandler.onDeleted(Collections.singletonList(new FlinkPod(flinkConfig, pod)));
+						break;
+					default:
+						LOG.info("Skip handling {} event for pod {}", action, pod.getMetadata().getName());
+						break;
+				}
+			}
+
+			@Override
+			public void onClose(KubernetesClientException e) {
+				LOG.error("Pods watcher onClose");
+				if (e != null) {
+					LOG.error(e.getMessage(), e);
+				}
+			}
+		};
+		this.internalClient.pods().withLabels(labels).watch(watcher);
+	}
+
+	@Override
+	public void close() {
+		this.internalClient.close();
+	}
+
+	// Private util methods
+
+	private CompletableFuture<FlinkService> createService(String serviceName,
+			List<Decorator<Service, FlinkService>> serviceDecorators) {
+		FlinkService flinkService = new FlinkService(this.flinkConfig);
+		for (Decorator<Service, FlinkService> d : serviceDecorators) {
+			flinkService = d.decorate(flinkService);
+		}
+		LOG.info("Create service with spec: {}", flinkService.getInternalResource().getSpec().toString());
+		this.internalClient.services().create(flinkService.getInternalResource());
+
+		ActionWatcher<Service> watcher = new ActionWatcher<>(Watcher.Action.ADDED,
+			flinkService.getInternalResource());
+		Watch watchConnectionManager = this.internalClient.services()
+			.inNamespace(this.nameSpace).withName(serviceName).watch(watcher);
+
+		return CompletableFuture.supplyAsync(() -> {
+			Service createdService = watcher.await(1, TimeUnit.MINUTES);
```

Comment:
maybe it'll be better if you add more comments about why CompletableFuture should be used.

Reply:
I have added comment in `FlinkKubeClient`. A CompletableFuture is returned by `createInternalService` and could used to wait for service ready.